eng/scripts/Generate-ServiceDirectories-From-Project-List.ps1

```diff
@@ -0,0 +1,116 @@
+    $xmlPomFile = $null
+    $xmlPomFile = New-Object xml
+    $xmlPomFile.Load($file.FullName)
+    $library = $xmlPomFile.project.groupId + ":" + $xmlPomFile.project.artifactId
+    $serviceDirectory = (Get-Item $file).Directory.Parent
+    # This if check is only necessary because resourcemanager and resourcemanagerhybrid contain the
+    # exact same group/artifact ids
+    if ($file.FullName.Split([IO.Path]::DirectorySeparatorChar) -notcontains "resourcemanagerhybrid") {
+        # The directories for sparse checkout and ServiceDirectories need the $SourcesDirectory
+        # stripped off
+        $tempDir = $serviceDirectory.FullName.Replace("$SourcesDirectory", "")
+        $tempDir = $tempDir.Replace([IO.Path]::DirectorySeparatorChar, '/')
+        $sparseCheckoutDirHash.Add($library, $tempDir)
+        if ($tempDir.StartsWith("/sdk/")) {
+            # Strip off the "/sdk/" to get the service directory. A ServiceDirectory will always have
+            # the format of /SDK/<ServiceDirectory>. Whereas sparse checkout can have other directories
+            # outside of /SDK, eg. /common
+            $tempDir = $tempDir.Replace("/sdk/", "")
+            $serviceDirHash.Add($library, $tempDir)
+        }
+    }
+}
+
+$sparseCheckoutDirectories = @()
+$serviceDirectories = @()
+$sparseCheckoutDirectories += "/sdk/parents"
+foreach ($project in $ProjectList) {
+    if ($sparseCheckoutDirHash.ContainsKey($project)) {
+        $sparseCheckoutDirectories += $sparseCheckoutDirHash[$project]
+        $serviceDirectories += $serviceDirHash[$project]
+    } else {
+        LogError("Did not find $project in any pom files")
+        $script:FoundError = $true
+    }
+}
+# Unreleased_ libraries are special. They're the only case, outside of FromSource runs where
+# libraries from other service directories, outside of the one we're building, need to get built.
+# It's easier to just add any unreleased_ dependency service directories to the sparse checkout
+# and service directory variables. The alternative requires a lot of specific pom file and
+# dependent pom file processing which is messy. The result here is that we may have an extra
+# service directory being sunk up when it wasn't necessary. The overall time of that sync
+# is a measured in seconds which isn't expensive and far outweighs the alternative mentioned
+# above.
+foreach ($project in $unreleasedList) {
+    if ($sparseCheckoutDirHash.ContainsKey($project)) {
+        $sparseCheckoutDirectories += $sparseCheckoutDirHash[$project]
+    } else {
+        LogError("Did not find unreleased $project in any pom files")
+        $script:FoundError = $true
+    }
+}
+
+if ($script:FoundError) {
+    exit 1
+}
+
+$temp =  ConvertTo-Json @($sparseCheckoutDirectories | Sort-Object | Get-Unique) -Compress
+Write-Host "setting env variable SparseCheckoutDirectories = $temp"
+Write-Host "##vso[task.setvariable variable=SparseCheckoutDirectories;]$temp"
```

Comment:
I'm not sure what the limit is but I suspect there is a limit on the size of a DevOps variable or a Env variable. So might be something we need to think about.

Reply:
@Wes - Previously the sparse checkout was granular down to the artifact's directory in some cases there were 120+ items added to the sparse checkout (like core's FromSource run for example). I changed things to just simply do the service directory which turned 120+ into like 30. If there's a limit it would have been hit long before I'd made my changes which dramatically trimmed things down.